multinet/__init__.py

```diff
@@ -21,10 +21,21 @@
 sentry_sdk.init(dsn=sentry_dsn, integrations=[FlaskIntegration()])
 
 
+def get_cors_hosts() -> List[str]:
+    """Read in comma-separated CORS hosts list from environment."""
+    cors_hosts = os.getenv("CORS_HOSTS", default=None)
+    if cors_hosts is None:
+        return []
+
+    return cors_hosts.split(",")
+
+
 def create_app(config: Optional[MutableMapping] = None) -> Flask:
     """Create a Multinet app instance."""
     app = Flask(__name__)
-    CORS(app)
+
+    cors_hosts = get_cors_hosts()
+    CORS(app, origins=cors_hosts, supports_credentials=True)
```

Comment:
I don't think it changes public vs. private. It should just mean that if we need to use multinet cookies from queries from another site, we can. So when you're on the nodelink, if you need to access a private workspace, the api server will check the credentials before returning it. If you're accessing a public workspace, we shouldn't have to check the cookie so it shouldn't matter

Reply:
Actually, @AlmightyYakob is right and I should have mentioned this in the PR description (I will update it for completeness).

It dawned on me that it won't make sense to have such a separation, at least not without really thinking about it. Almost all of our endpoints specify a workspace, and already you will run into login/permissions issues, which requires cookies. So I went ahead and switched over the whole API to require a whitelist entry for any connecting hosts.

What I _would_ like to do is allow non-cookie connections from `*`, but I don't think that's possible to do via the Flask CORS thing. If I am missing something, let me know.

In the meantime, I think it's ok if we revert to not allowing just anyone to connect to the main Multinet server. For now, we'll essentially have to be like Google, and solicit requests to connect to our Multinet instance which we can respond to by adding a host to the whitelist and going from there. Since this is a research project in progress, I think this is an ok situation for now.

Reply:
This [tutorial](https://dev.to/g33konaut/understanding-cors-aaf) shows how server logic can decide whether to accept a non-credentialed request from anyone, and allow credentialed requests from a whitelist. To me, that means that if flask-cors doesn't allow this, we should seek out a solution that does (up to and including creating our own middleware to accomplish this). But that would be in a followup issue.

Reply:
Oh, of course. My mistake. This will cause issues for us with local testing of the vis apps as they currently stand. They make requests to multinet.app to grab live data when testing locally. You can specify a `host` in the querystring but that would require setting up the whole infrastructure locally for vis app testing. Not that it's necessarily a bad thing, just something to consider.

Reply:
If we can specify `host` in the querystring, can we not just point that to a local Multinet instance?

Reply:
@waxlamp I agree with your short term solution. Regarding splitting up authenticated vs non-authenticated requests in the future, it seemed like this was possible through Flask CORS with the `resources` field. If that's not the case, we still have an out through the flask [`after_request`](https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.after_request) hook, where we could just check the domain manually and set any headers we want in the response object.

Reply:
I opened #384 to track these ideas.